Fail tests on warning [Priority Needed]

anshgoyalevil opened the pull request:
<!--
This section can be deleted after reading.

We employ the following branching strategy to simplify the development process and to ensure that only stable code is pushed to the `master` branch:

- `develop`: For unstable code: New features and bug fixes.
- `alpha-x.x.x`: For stability testing: Only bug fixes accepted.
- `master`: Where the stable production ready code lies. Only security related bugs.

-->

<!--
Thanks for submitting a pull request! Please provide enough information so that others can review your pull request.
-->

**What kind of change does this PR introduce?**
This PR fixes the following things:
- Fail tests on warnings
- Fix existing warnings

<!-- E.g. a bugfix, feature, refactoring, etc… -->

**Issue Number:**

Fixes #819 

**Did you add tests for your changes?**
Yes
<!--Yes or No. Note: Add unit tests or automation tests for your code.-->

**Snapshots/Videos:**
NR
<!--Add snapshots or videos wherever possible.-->

**If relevant, did you update the documentation?**
Not Relevant
<!--Add link to Talawa-Docs.-->

**Summary**
This PR fixes the most important problem of warning being pushed by contributors, which were conflicting with workflows.
<!-- Explain the **motivation** for making this change. What existing problem does the pull request solve? -->
<!-- Try to link to an open issue for more information. -->

**Does this PR introduce a breaking change?**
No
<!-- If this PR introduces a breaking change, please describe the impact and a migration path for existing applications. -->

**Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa-admin/blob/master/CONTRIBUTING.md)?**
Yes
<!--Yes or No-->

anshgoyalevil commented:
There is a test which is Workflow specific. Working on it.

palisadoes commented:
Won't we need to adjust eslint as well?

anshgoyalevil commented:
@palisadoes 

No, That thing is done by setupTests.ts

I am working on the tests.

anshgoyalevil commented:
@palisadoes 

This PR is ready to be merged.

Here's an explanation of what I have done here:

- React Testing Library and Jest automatically looks for a file named ```setupTests.ts``` which contain the global configurations for tests.
- Added the option where on receiving any warning or error, the console would give error. 
- Commented out the code for ```AddOnRegister``` component because it was completely dependent on Plugins system, thus working with it was not possible.
- Used the ```StaticMockLink``` helper function which I created previously in another merged PR.

palisadoes commented:
1. @anshgoyalevil Will the tests giving warnings fail in all future PRs now?
2. @tasneemkoushar  Please review

anshgoyalevil commented:
@palisadoes 

All the tests which give warning from now on will be failed.

palisadoes commented:
We don't want to add features because of duplicating those in GSoC ideas.

This PR removes existing code that will have to be replaced by a GSoC idea.

Are you sure you have investigated this properly? The aim was to make sure we test all aspects properly. Removing code in this case seems short sighted.

anshgoyalevil commented:
@palisadoes Sir, I haven't removed the code. I have commented it because it has to be redesigned during the GSoC ideas.

There is no link between this issue and that code. If you would suggest, I would uncomment that code. It wouldn't affect the motive of this issue and PR.

palisadoes commented:
So it's just the tests for the code? I was to hasty.

anshgoyalevil commented:
No, I have commented the tests (```AddOnRegister.test.tsx```) as well as code for ```AddOnRegister.tsx```

This is because the tests were testing a hypothetical page.

And, if I just have commented out those tests, the code coverage would have declined.

If there's no problem with code coverage being declined, I would uncomment the code in ```AddOnRegister.tsx``` file.

anshgoyalevil commented:
No, I have commented the tests (```AddOnRegister.test.tsx```) as well as code for ```AddOnRegister.tsx```

This is because the tests were testing a hypothetical page.

And, if I just have commented out those tests, the code coverage would have declined.

If there's no problem with code coverage being declined, I would uncomment the code in ```AddOnRegister.tsx``` file.

anshgoyalevil commented:
Coverage works in a way that, if we call a function inside a test, the code coverage would rise, no matter that function's expected value is matched or not.

palisadoes commented:
Thanks for the clarification

SiddheshKukade commented on src/components/AddOn/core/AddOnRegister/AddOnRegister.tsx:
```diff
@@ -1,158 +1,166 @@
-                required
-                value={formState.pluginName}
-                onChange={(e) => {
-                  setFormState({
-                    ...formState,
-                    pluginName: e.target.value,
-                  });
-                }}
-              />
-            </Form.Group>
-            <Form.Group className="mb-3" controlId="registerForm.PluginName">
-              <Form.Label>{t('creatorName')}</Form.Label>
-              <Form.Control
-                type="text"
-                placeholder={t('cName')}
-                autoComplete="off"
-                required
-                value={formState.pluginCreatedBy}
-                onChange={(e) => {
-                  setFormState({
-                    ...formState,
-                    pluginCreatedBy: e.target.value,
-                  });
-                }}
-              />
-            </Form.Group>
-            <Form.Group className="mb-3" controlId="registerForm.PluginURL">
-              <Form.Label>{t('pluginDesc')}</Form.Label>
-              <Form.Control
-                // type="text"
-                rows={3}
-                as="textarea"
-                placeholder={t('pDesc')}
-                required
-                value={formState.pluginDesc}
-                onChange={(e) => {
-                  setFormState({
-                    ...formState,
-                    pluginDesc: e.target.value,
-                  });
```
hello @anshgoyalevil, I'd like to know why have you commented on this code and what tests were failing for this. the test for this only contains some UI checks to make sure proper UI is rendered for the plugin. Commenting on this code would remove that UI from admin.
Also  how this helps to solve #819 as it was just intended to implement a check that tests should fail when there's a warning in the console for that test case, right?

anshgoyalevil replied:
@SiddheshKukade 

Thanks for asking out.
This code was for adding the plugin, and the plugins system doesn't work. With this code in place, the coverage would fall.

I have a question too. Should we test a component, if it doesn't work? Are we aiming for coverage or for features?

Moreover, if you think, there was no problem with the code, since you are also having some valid points. I would add the code back. But, I don't think the implementation is correct. Plugins are not posts. They are predefined, and not taken as a user input.

SiddheshKukade commented:
> If there wouldn't be a plugins page which could add the plugin, how can we test that. And if we would try doing that, we would get errors and warnings, like previously.

Hello @anshgoyalevil, just for reference, the current plugin store and their linking with the Talwa mobile was working fine, but after some time this PR : https://github.com/PalisadoesFoundation/talawa/pull/1482 delete code containing some important logic  and detailed description from it, I think, due to unavailable documentation of the file. I am working on a adding it back

SiddheshKukade commented:
> If there wouldn't be a plugin page which could add the plugin, how can we test that. And if we would try doing that, we would get errors and warnings, like previously.

About adding a plugin: 
There is a button on the plugin page `+ Add new`  to add new plugins to the store.

Hello @anshgoyalevil, just for reference, the current plugin store and their linking with the Talwa mobile was working fine, but after some time this PR: https://github.com/PalisadoesFoundation/talawa/pull/1482 delete code containing some important logic and detailed description from it.
I am working on adding it back to make sure that It works fine again. 

The current GSoC project idea was actually on optimizing this process (it was working before you can refer to the videos on youtube) 

I was having this week which would finish, next Tuesday and I would also get started to work on this from this weekend.

SiddheshKukade commented:
> No, I have commented the tests (`AddOnRegister.test.tsx`) as well as code for `AddOnRegister.tsx`
> 
> This is because the tests were testing a hypothetical page.

I've explained here about why it is not hypothetical, https://github.com/PalisadoesFoundation/talawa-admin/pull/856#issuecomment-1500911234 


> And, if I just have commented out those tests, the code coverage would have declined.
> 
> If there's no problem with code coverage being declined, I would uncomment the code in `AddOnRegister.tsx` file.

I think the tests for the `AddOnRegister.tsx` were passing as it wasn't modified after I created that. 

we can also work with less coverage for this pull request, @palisadoes sir is that okay?

anshgoyalevil commented:
@SiddheshKukade Also, the plugin system is incorrect in whole.

We are taking the input of a plugin in a form, as if its a post or event.

Plugins are pre-defined and not taken as an input, and if taken, they need to contain a file, like how WordPress works.

I cannot comment on the complete plugins architecture, but as much I know, the plugins page is totally messed up.

Also, the mobile implementation of plugins is very inefficient. We are aiming for orgs with low budget, and the current Talawa Mobile plugins function sends over 20 requests per second. Let me send you a graph to understand it better.

![image](https://user-images.githubusercontent.com/94157520/230729420-7145fda0-8eb6-46b8-80bb-042cbe729f27.png)


Total requests sent to the API over 1 months is 2.56 Million. With only about 50-60 users. It could be less than 100K, if we would have an efficient plugins system.

I wouldn't go on comparing the opex and capex for small orgs, like the aim of Talawa and Palisadoes.

I would revert that commented out code in this PR. Hope you find this suggestion useful.

anshgoyalevil commented:
@SiddheshKukade 

Would you please guide me in solving this error. 

This is occurring in the unchanged code which was previously written by you.

```
Testing AddOnRegister › should render modal and take info to add plugin for registered organization

    TestingLibraryElementError: Unable to find an element with the placeholder text of: /pName/i

    <body
      class="modal-open"
      style="padding-right: 0px;"
    >
      <div
        aria-hidden="true"
      >
        <button
          class="modalbtn btn btn-primary"
          type="button"
        >
          <i
            class="fa fa-plus"
          />
          Add New
        </button>
      </div>
      <div
        class="fade modal-backdrop show"
      />
      <div
        aria-modal="true"
        class="fade modal show"
        role="dialog"
        style="display: block;"
        tabindex="-1"
      >
        <div
          class="modal-dialog"
        >
          <div
            class="modal-content"
          >
            <div
              class="modal-header"
            >
              <div
                class="modal-title h4"
              >

                Add Plugin
              </div>
              <button
                class="close"
                type="button"
              >
                <span
                  aria-hidden="true"
                >
                  ×
                </span>
                <span
                  class="sr-only"
                >
                  Close
                </span>
              </button>
            </div>
            <div
              class="modal-body"
            >
              <form
                class=""
              >
                <div
                  class="mb-3 form-group"
                >
                  <label
                    class="form-label"
                    for="registerForm.PluginName"
                  >
                    Plugin Name
                  </label>
                  <input
                    autocomplete="off"
                    class="form-control"
                    id="registerForm.PluginName"
                    placeholder="Ex: Donations"
                    required=""
                    type="text"
                    value=""
                  />
                </div>
                <div
                  class="mb-3 form-group"
                >
                  <label
                    class="form-label"
                    for="registerForm.PluginName"
                  >
                    Creator Name
                  </label>
                  <input
                    autocomplete="off"
                    class="form-control"
                    id="registerForm.PluginName"
                    placeholder="Ex: john Doe"
                    required=""
                    type="text"
                    value=""
                  />
                </div>
                <div
                  class="mb-3 form-group"
                >
                  <label
                    class="form-label"
                    for="registerForm.PluginURL"
                  >
                    Plugin Description
                  </label>
                  <textarea
                    class="form-control"
                    id="registerForm.PluginURL"
                    placeholder="This Plugin enables UI for"
                    required=""
                    rows="3"
                  />
                </div>
              </form>
            </div>
            <div
              class="modal-footer"
            >
              <button
                class="btn btn-secondary"
                data-testid="addonclose"
                type="button"
              >
                Close
              </button>
              <button
                class="btn btn-primary"
                data-testid="addonregister"
                type="button"
              >
                Register
              </button>
            </div>
          </div>
        </div>
      </div>
    </body>

      51 |
      52 |     userEvent.click(screen.getByRole('button', { name: /addnew/i }));
    > 53 |     userEvent.type(screen.getByPlaceholderText(/pName/i), 'myplugin');
         |                           ^
      54 |     userEvent.type(screen.getByPlaceholderText(/pDesc/i), 'test description');
      55 |     userEvent.type(screen.getByPlaceholderText(/cName/i), 'test creator');
      56 |     userEvent.click(screen.getByTestId('addonregister'));

      at Object.getElementError (node_modules/@testing-library/dom/dist/config.js:37:19)
      at node_modules/@testing-library/dom/dist/query-helpers.js:90:38
      at node_modules/@testing-library/dom/dist/query-helpers.js:62:17
      at getByPlaceholderText (node_modules/@testing-library/dom/dist/query-helpers.js:111:19)
      at Object.<anonymous> (src/components/AddOn/core/AddOnRegister/AddOnRegister.test.tsx:53:27)

```

anshgoyalevil commented:
@palisadoes Sir, Done with the changes. I have restored the ```AddOnRegister.tsx``` and ```AddOnRegister.test.tsx``` file on @SiddheshKukade opinion.

Also, fixed the tests in ```AddOnRegister.test.tsx``` which were now failing (Previously giving warnings).

SiddheshKukade commented:
> @SiddheshKukade Also, the plugin system is incorrect in a whole.

@anshgoyalevil  Again, It was working fine as it should be for all the UI feature and the youtube videos is also available as working proof. 
The entire architecture was working as we've planned it last year. 

The reason for the current inconsistent state about plugins due to merge of PR's like https://github.com/PalisadoesFoundation/talawa/pull/1482 in which contributors haven't read this https://docs.talawa.io/docs/developers/talawa-admin/plugins/plugin-architecture guide in which I've explained which files are important and which files are working in what sense. They deleted those lines of code without consideration and I'm now working on fixing them to make it work again as it was previously.

> We are taking the input of a plugin in a form as if it's a post or event.
> Plugins are pre-defined and not taken as input, and if taken, they need to contain a file, like how WordPress works.

The creation of the plugin is only allowed to **SUPERADMINS**. WordPress I think is not a good option to compare here as it's a completely different tool and here we're discussing about controlling a mobile app from a web portal. 

> I cannot comment on the complete plugin architecture, but as much I know, the plugins page is totally messed up.

You can explain about what do you feel as _**messed up**_ here ?
I'll also suggest to take a look at https://docs.talawa.io/docs/developers/talawa-admin/plugins/
 
> Also, the mobile implementation of plugins is very inefficient. We are aiming for orgs with low budget, and the current Talawa Mobile plugins function sends over 20 requests per second. Let me send you a graph to understand it better.

I know about that, in the last year focus was on **providing an instantaneous feedback to users based on the feature changes,** I considered the talawa app to be completely self-hosted on private hardware instead of being a SaaS, but I can see the change in direction this year.

Hope this helps and clears your doubts.

SiddheshKukade commented:
>I have restored the `AddOnRegister.tsx` and `AddOnRegister.test.tsx` file on @SiddheshKukade opinion.
> Also, fixed the tests in `AddOnRegister.test.tsx` which were now failing (Previously giving warnings).

@anshgoyalevil tests looks, fine here
![image](https://user-images.githubusercontent.com/65951872/230734699-6c38c416-385d-4d15-8cca-d35a9b11c2e5.png)

anshgoyalevil commented:
> @anshgoyalevil Again, It was working fine as it should be for all the UI feature and the youtube videos is also available as working proof.
The entire architecture was working as we've planned it last year.

Yes. Agreed.

> The creation of the plugin is only allowed to SUPERADMINS. WordPress I think is not a good option to compare here as it's a completely different tool and here we're discussing about controlling a mobile app from a web portal.

I guess you didn't get my point. Let's take the example of BreezeChms. It offers pre-defined plugins like ```Mailchimp```, etc. 

What is the point of that ```Add New``` button?

Let's go back in time, when it was working.

- I clicked ```Add New```
- I am a SUPERADMIN, so I entered a plugin name and description.
- Now, what would happen next? From where would the plugin code come, by adding it from dashboard.
- The plugins like ```Events```, ```Chats```, ```Posts``` are working because their implementation is present is codebase, but thats not the case with a new plugin added from dashboard via the ```Add New``` button.

> You can explain about what do you feel as messed up here ?
I'll also suggest to take a look at https://docs.talawa.io/docs/developers/talawa-admin/plugins/

This link is not working.

> I know about that, in the last year focus was on providing an instantaneous feedback to users based on the feature changes, I considered the talawa app to be completely self-hosted on private hardware instead of being a SaaS, but I can see the change in direction this year.

Fine.

-------------------------

The implementation of Plugin Architecture is a debatable topic. We can continue the conversation on slack, if you have any query.

Thanks again for explaining in detail.

literalEval commented:
> > If there wouldn't be a plugin page which could add the plugin, how can we test that. And if we would try doing that, we would get errors and warnings, like previously.
> 
> About adding a plugin: 
> There is a button on the plugin page `+ Add new`  to add new plugins to the store.
> 
> Hello @anshgoyalevil, just for reference, the current plugin store and their linking with the Talwa mobile was working fine, but after some time this PR: https://github.com/PalisadoesFoundation/talawa/pull/1482 deleted code containing some important logic and detailed description from it.
> I am working on adding it back to make sure that It works fine again. 
> 
> The current GSoC project idea was actually on optimizing this process (it was working before you can refer to the videos on youtube) 
> 
> I was having written exams this week which would finish next tuesday and I would also get started to work on this from this weekend. 

The deletion was only done after having a detailed discussion with other volunteers. The app was unnecessarily polling the server, and had a very bad architecture in whole. I hope it get fixed this year by a actually well thought architecture.

SiddheshKukade commented:
> I guess you didn't get my point. Let's take the example of BreezeChms. It offers pre-defined plugins like `Mailchimp`, etc.

The mailchimp plugin on the breezechm is working for providing API integration on the website. It doesn't control any mobile UI


> What is the point of that `Add New` button?
> Let's go back in time, when it was working.
>     * I clicked `Add New`
>     * I am a SUPERADMIN, so I entered a plugin name and description.
>     * Now, what would happen next? From where would the plugin code come, by adding it from dashboard.
>     * The plugins like `Events`, `Chats`, `Posts` are working because their implementation is present is codebase, but thats not the case with a new plugin added from dashboard via the `Add New` button.
SUPERADMIN are developers or owners of the product who can have control over all orgs. Hence they hold the right to publish new feature as a plugin when it is implemented.

> This link is not working.
 Here's another one : https://docs.talawa.io/docs/category/plugins

anshgoyalevil commented:
@SiddheshKukade 

Thanks for clarifying the points.

> The mailchimp plugin on the breezechm is working for providing API integration on the website. It doesn't control any mobile UI
> SUPERADMIN are developers or owners of the product who can have control over all orgs. Hence they hold the right to publish new feature as a plugin when it is implemented.

You are still not getting my concerns. You can ping me over slack. We can discuss it there. With more discussions, better ideas would churn out, and better would be the end product.

> Here's another one : https://docs.talawa.io/docs/category/plugins

Thanks again for it.

anshgoyalevil commented:
@SiddheshKukade 

Thanks for clarifying the points.

> The mailchimp plugin on the breezechm is working for providing API integration on the website. It doesn't control any mobile UI
> SUPERADMIN are developers or owners of the product who can have control over all orgs. Hence they hold the right to publish new feature as a plugin when it is implemented.

You are still not getting my concerns. You can ping me over slack. We can discuss it there. With more discussions, better ideas would churn out, and better would be the end product.

> Here's another one : https://docs.talawa.io/docs/category/plugins

Thanks again for it.

palisadoes commented:
We have two issues here:
1. The desire to have tests fail on warnings
1. The need to preserve plugin functionality as originally implemented.

Does this PR solve the first concern while not affecting the current state of plugins?

Further rectification of the plugin system needs to be separate. Can we open a discussion on this in the admin slack channel?

anshgoyalevil commented:
@palisadoes 

Sir, it follows both of these point.

1. Tests will now fail on warning
2. The Plugins functionality is preserved.

I think, you are not sure about the first point, since all tests are passing.
The reason for that is, I have fixed over 20 tests which were failing after I added the functionality to fail tests on warning.

anshgoyalevil commented:
The other doubt you might have is, I have not changed the eslint.json or tsconfig.json, then how it will fail tests on warnings.

The answer to that is, we are using the setupTests.ts file to extend this functionality.

Checkout that file in this PR. It clearly shows how errors are asked to be thrown on recieving ```global.console.warning```

palisadoes commented:
What are we doing about restoring the plugin functionality?

anshgoyalevil commented:
> What are we doing about restoring the plugin functionality?

The files related to Plugin Functionality, like ```AddOnRegister.tsx``` are not affected in this PR. I commented out the code earlier, but un-commented it after @SiddheshKukade asked to do so.

If you are asking about how would it work, as it is not in working state from last few months.

I think, @SiddheshKukade is working on it. Would be happy to get involved in re-enabling that feature.

SiddheshKukade commented:
> The changes in that PR were only done after having a detailed discussion with other volunteers. The app was unnecessarily polling the server, and had a very bad architecture in whole. I hope it get fixed this year by an actually well thought architecture. 
> 
I've explained in the previous message why it's designed this way and It was intended for the last year goal for talawa being completely self hosted.

> Moreover, there's no "deletion" in that PR if you look closely. 

If you look at the history of the commits there was some deletion in the addOnRegister code.

SiddheshKukade commented:
> What are we doing about restoring the plugin functionality?

I'm working on that will get started from today. I need to figure out which first which part's have been deleted and then will work on adding them back.

palisadoes commented:
Thanks for working on restoring the functionality.

palisadoes commented:
Please create an issue for it @SiddheshKukade

literalEval commented:
@SiddheshKukade

> @literalEval I've explained in the previous message why it's designed this way and It was intended for the last year goal for talawa being completely self hosted.

There were far better ways for that to be implemented

> If you look at the history of the commits there was some deletion in the addOnRegister code.

I have never worked on Talawa Admin, so I have no idea what you are talking about. Someone who has looked at the admin code closely will answer this better.

> @literalEval And Also I think you didn't read the docs before deleting the code in https://github.com/PalisadoesFoundation/talawa/pull/1482 I've explained there that the file was important.

The changes which I did followed this [PR](https://github.com/PalisadoesFoundation/talawa/pull/1469). Your implementation caused a lot of `KeyError` and `DDoS attack simulations`. Moreover, the PR you mentioned merely made some things non-plugins, as discussed with the volunteers. 

@palisadoes THE "PLUGIN SYSTEM" IS NOT BROKEN in the app side. I don't know about the admin side, but the app side still perfectly works and needs no change. The only thing that I did was to make some things as non-plugins and available to everyone, and that too because I was not provided with the list of what should be a plugin and what should not be. I'll quickly make them plugins again. THERE'S NO CHANGE NEEDED IN THE APP SIDE.

palisadoes commented:
1. Please sort this out in the newly created respective issues. This PR is already merged.
2. @literalEval @SiddheshKukade refer to the files in the issues to state your case. 
3. Please keep the discussion professional. The general tone is not beneficial to our community of collaboration.

literalEval commented:
@palisadoes understood sir. We'll discuss together and conclude a better approach to this issue :+1:

SiddheshKukade commented:
> I have never worked on Talawa Admin, so I have no idea what you are talking about. Someone who has looked at the admin code closely will answer this better.

@literalEval Then,  I would suggest to take a look at it first. Using all apps at once is the best ways to visualize plugins

I've already answered all of your other  concern in previous replies in depth.
